Tests/NetworkServiceTests/NetworkServiceTests.swift

```diff
@@ -1,8 +1,170 @@
+        // Given
+        MockURLProtocol.requestHandler = { request in
+            let response = try XCTUnwrap(HTTPURLResponse(url: XCTUnwrap(request.url), statusCode: 500, httpVersion: nil, headerFields: nil))
+            return (response, nil)
+        }
+        
+        let endpoint = MockEndpoint()
+        
+        // When
+        do {
+            _ = try await networkService.request(endpoint)
+            XCTFail("Expected failure, but got success")
+        } catch let error as NetworkError {
+            XCTAssertEqual(error, .serverError)
+        } catch {
+            XCTFail("Expected NetworkError.badRequest, but got \(error)")
+        }
+    }
+    
+    func testUnknownErrorStatusCodeInResponseTriggersUnknownError() async throws {
+        // Given
+        MockURLProtocol.requestHandler = { request in
+            let response = try XCTUnwrap(HTTPURLResponse(url: XCTUnwrap(request.url), statusCode: 300, httpVersion: nil, headerFields: nil))
+            return (response, nil)
+        }
+        
+        let endpoint = MockEndpoint()
+        
+        // When
+        do {
+            _ = try await networkService.request(endpoint)
+            XCTFail("Expected failure, but got success")
+        } catch let error as NetworkError {
+            XCTAssertEqual(error, .unknownError)
+        } catch {
+            XCTFail("Expected NetworkError.badRequest, but got \(error)")
+        }
+    }
+    
+    func testRequestWithURLParametersReturnsExpectedData() async throws {
+        // Given
+        let expectedData = try XCTUnwrap("Mocked Response Data with URL Parameters".data(using: .utf8))
+        MockURLProtocol.requestHandler = { request in
+            let response = try XCTUnwrap(HTTPURLResponse(url: XCTUnwrap(request.url), statusCode: 200, httpVersion: nil, headerFields: nil))
+            return (response, expectedData)
+        }
+
+        var endpoint = MockEndpoint()
+        endpoint.task = .requestWithURLParameters(urlParameters: ["key": "value"])
+
+        // When
+        let data = try await networkService.request(endpoint)
+
+        // Then
+        XCTAssertEqual(data, expectedData)
+    }
+    
+    func testRequestWithBodyParametersReturnsExpectedData() async throws {
+        // Given
+        let expectedData = try XCTUnwrap("Mocked Response Data with Body Parameters".data(using: .utf8))
+        MockURLProtocol.requestHandler = { request in
+            let response = try XCTUnwrap(HTTPURLResponse(url: XCTUnwrap(request.url), statusCode: 200, httpVersion: nil, headerFields: nil))
+            return (response, expectedData)
+        }
+        
+        var endpoint = MockEndpoint()
+        endpoint.task = .requestWithBodyParameters(bodyParameters: ["key": "value"])
+        
+        // When
+        let data = try await networkService.request(endpoint)
+        
+        // Then
+        XCTAssertEqual(data, expectedData)
     }
 }
```

Comment:
We can create a variable `endpoint` and rewrite that in every iteration of `setUp` function. 
WDYT?
```suggestion
    }
    
        // MARK: - Private interface

    private var endpoint = MockEndpoint()
}
```

Reply:
I agree, fixed with 036eb7f